src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -156,39 +157,58 @@ module IncrementalBuildSyntaxTree =
                 fileParsed.Trigger fileName
 
                 input, sourceRange, fileName, diagnosticsLogger.GetDiagnostics()
+            )
 
+        let parse (parseTask: Threading.Tasks.Task<_>) =
+            try
+                parseTask.Result
             with exn ->
                 let msg = sprintf "unexpected failure in SyntaxTree.parse\nerror = %s" (exn.ToString())
                 System.Diagnostics.Debug.Assert(false, msg)
                 failwith msg
 
-        let parseOrSkip sigNameOpt =
+        // We control access to the cache to prevent theoretical possibility of starting the same parse job more than once.
+        // We lock only a single particular cache key, just to start the task, not for the duration of the task.
+        let getValue source = lock source <| fun () -> cache.GetValue(source, getParseTask)
+
+        /// Parse the given file and return the given input.
+        member _.Parse() =
             IncrementalBuilderEventTesting.MRU.Add(IncrementalBuilderEventTesting.IBEParsed fileName)
             use _ =
                 Activity.start "IncrementalBuildSyntaxTree.parseOrSkip"
                     [|
                         Activity.Tags.fileName, fileName                       
                         Activity.Tags.buildPhase, BuildPhase.Parse.ToString()
-                        "canSkip", (isImplFile && sigNameOpt |> Option.isSome).ToString()
                     |]
-
-            match  sigNameOpt with
-            | Some sigName when isImplFile -> parsedImplFileStub sigName
-            | _ when useCache ->
-                match cache.TryGetValue file with
+            if useCache then
+                match cache.TryGetValue source with
                 | true, result ->
                     Activity.addEvent Activity.Events.cacheHit
-                    result
-                | _ -> cache.GetValue(file, parse)
-            | _ -> parse file
+                    parse result
+                | _ -> cache.GetValue(source, getParseTask) |> parse
+            else
+                getParseTask source |> parse
 
-        /// Parse the given file and return the given input.
-        member _.Parse(sigNameOpt) = parseOrSkip sigNameOpt
-        
-        static member Invalidate(source) = cache.Remove(source) |> ignore
+        member _.HasSignature = hasSignature
+
+        member _.Skip = skippedImplFilePlaceholder
+      
+        member _.Invalidate() =
+            if useCache then
+                cache.Remove(source) |> ignore
+                getValue source |> ignore
```

Comment:
why do we need to call "getValue" during cache invalidation? This might start a new task, or?
(e.g. when calling Invalidate() multiple times)

Reply:
Yeah, repeated Invalidate would happen if the client spams with NotifyFileChanged. This can't be helped at the moment for two reasons: 
* we can't effectively cancel the parse.
* timestamp is not real versioning. Notification should come with version, like in Roslyn, (or just an `int` would do),  so we don't reparse needlessly.

Reply:
As for why, this is where the "eager" parsing happens.

Reply:
> Yeah, repeated parse would happen if the client spams with NotifyFileChanged. This can't be helped at the moment for two reasons: 
> * we can't effectively cancel the parse.
> * timestamp is not real versioning. Notification should come with version, like in Roslyn, (or just an `int` would do),  so we don't reparse needlessly.

Yeah, ideally we should be able to cancel parsing for the specific file, and have the ability to debounce requests, so we can always do it for the latest version.

Reply:
I will investigate how we can bring in cancellation.

What I would like to do even more (but I do not know how) would be common-prefix-aware debounce - i.e. cancelling and not starting from scratch, but starting from the last untouched token.

Reply:
I think it would be most beneficial to not parse when there are no changes. This is less of a problem with FileSystemShim, much more with live buffers in VS. VS spams a lot of workspace changed events, when a source didn't in fact change.

Reply:
I think this is only used in the service MRU cache, the background builder is notified just with a timestamp like this:
``` f#
            | Some builder -> do! builder.NotifyFileChanged(fileName, DateTime.UtcNow)
```
Which would be a concern if we wanted to parse eagerly and not lazily. Also, if we parse lazily, there is no guarantee of the source identity, when we do getSource.